temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowExecutor.java

```diff
@@ -36,13 +37,45 @@
 import io.temporal.internal.common.ProtobufTimeUtils;
 import io.temporal.internal.common.UpdateMessage;
 import io.temporal.internal.statemachines.WorkflowStateMachines;
+import io.temporal.internal.sync.SignalHandlerInfo;
+import io.temporal.internal.sync.UpdateHandlerInfo;
 import io.temporal.internal.worker.WorkflowExecutionException;
 import io.temporal.worker.MetricsType;
 import io.temporal.worker.NonDeterministicException;
+import io.temporal.workflow.HandlerUnfinishedPolicy;
+import java.util.List;
+import java.util.Map;
 import java.util.Optional;
+import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
 
 final class ReplayWorkflowExecutor {
+  @VisibleForTesting
+  public static final String unfinishedUpdateHandlesWarnMessage =
+      "[TMPRL1102] Workflow finished while update handlers are still running. This may "
+          + "have interrupted work that the update handler was doing, and the client "
+          + "that sent the update will receive a 'workflow execution already completed' "
+          + "Exception instead of the update result. You can wait for all update and "
+          + "signal handlers to complete by using `await workflow.Await(() -> workflow.isHandlersFinished())`. "
+          + "Alternatively, if both you and the clients sending the update are okay with "
+          + "interrupting running handlers when the workflow finishes, and causing "
+          + "clients to receive errors, then you can disable this warning via the update "
+          + "handler annotations: `@UpdateMethod(unfinishedPolicy = HandlerUnfinishedPolicy.ABANDON)`.";
+
+  @VisibleForTesting
+  public static final String unfinishedSignalHandlesWarnMessage =
+      "[TMPRL1102] Workflow finished while signal handlers are still running. This may "
+          + "have interrupted work that the update handler was doing. You can wait for all update and "
+          + "signal handlers to complete by using `await workflow.Await(() -> workflow.isHandlersFinished())`. "
+          + "Alternatively, if both you  are okay with "
+          + "interrupting running handlers when the workflow finishes, and causing "
+          + "clients to receive errors, then you can disable this warning via the update "
```

Comment:
```suggestion
          + "interrupting running handlers when the workflow finishes, "
          + "then you can disable this warning via the update "
```

I realized today that this part of the message doesn't apply to signals so have been going through removing it from each lang.

Reply:
It looks like "and the and the clients sending the signal" is missing:

> `if both you  are okay with`